File: src/test/java/by/architecture_map/belarus/service/impl/ArchitecturalStyleServiceImplTest.java
package by.architecture_map.belarus.service.impl;

import by.architecture_map.belarus.entity.ArchitecturalStyle;
import by.architecture_map.belarus.exception.NotFoundException;
import by.architecture_map.belarus.service.ImageService;
import by.architecture_map.belarus.repository.jpa.ArchitecturalStyleRepository;
import org.junit.jupiter.api.Test;

import org.junit.jupiter.api.Assertions.*;

import java.util.*;

import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class ArchitecturalStyleServiceImplTest {

    @Mock
    private ArchitecturalStyleRepository architecturalStyleRepository;

    @Mock
    private ImageService imageService;

    @InjectMocks
    private ArchitecturalStyleServiceImpl architecturalStyleService;

    @Test
    void whenSaveArchitecturalStyle_thenSaveStyle() {
        // given
        ArchitecturalStyle architecturalStyle = new ArchitecturalStyle();
        architecturalStyle.setName("Gothic");
        architecturalStyle.setId(1);
        when(architecturalStyleRepository.save(architecturalStyle)).thenReturn(architecturalStyle);

        // when
        ArchitecturalStyle result = architecturalStyleService.create(architecturalStyle);

        // then
        verify(architecturalStyleRepository, times(1)).save(architecturalStyle);
        assertEquals(architecturalStyle, result);
    }

    @Test
    void whenFindAllArchitecturalStyles_thenReturnListOfStyles() {
        // given
        ArchitecturalStyle gothic = new ArchitecturalStyle();
        gothic.setName("Gothic");
        gothic.setId(1);
        ArchitecturalStyle baroque = new ArchitecturalStyle();
        baroque.setName("Baroque");
        baroque.setId(2);
        List<ArchitecturalStyle> styles = Arrays.asList(gothic, baroque);
        when(architecturalStyleRepository.findAll()).thenReturn(styles);

        // when
        List<ArchitecturalStyle> result = architecturalStyleService.findAll();

        // then
        verify(architecturalStyleRepository, times(1)).findAll();
        assertEquals(styles, result);
    }

    @Test
    void whenFindArchitecturalStyle_thenReturnStyle() {
        // given
        int id = 1;
        ArchitecturalStyle architecturalStyle = new ArchitecturalStyle();
        architecturalStyle.setName("Gothic");
        architecturalStyle.setId(id);
        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.of(architecturalStyle));

        // when
        ArchitecturalStyle result = architecturalStyleService.find(id);

        // then
        verify(architecturalStyleRepository, times(1)).findById(id);
        assertEquals(architecturalStyle, result);
    }

    @Test
    void whenFindArchitecturalStyleAndStyleDoesNotExist_thenThrowNotFoundException() {
        // given
        int id = 1;
        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.empty());

        // when & then
        assertThrows(NotFoundException.class, () -> architecturalStyleService.find(id));

        // verify
        verify(architecturalStyleRepository, times(1)).findById(id);
    }

    @Test
    void whenUpdateArchitecturalStyle_thenUpdateStyle() {
        // given
        int id = 1;
        ArchitecturalStyle existingStyle = new ArchitecturalStyle();
        existingStyle.setName("Gothic");
        existingStyle.setId(id);
        ArchitecturalStyle updatedStyle = new ArchitecturalStyle();
        updatedStyle.setName("Updated Gothic");
        updatedStyle.setId(id);

        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.of(existingStyle));
        when(architecturalStyleRepository.save(existingStyle)).thenAnswer(invocation -> {
            existingStyle.setName(updatedStyle.getName());
            return existingStyle;
        });

        // when
        ArchitecturalStyle result = architecturalStyleService.update(id, updatedStyle);

        // then
        verify(architecturalStyleRepository, times(1)).findById(id);
        verify(architecturalStyleRepository, times(1)).save(existingStyle);
        assertEquals(updatedStyle.getName(), result.getName());
    }

    @Test
    void whenUpdateArchitecturalStyleAndStyleDoesNotExist_thenThrowNotFoundException() {
        // given
        int id = 1;
        ArchitecturalStyle updatedStyle = new ArchitecturalStyle();
        updatedStyle.setName("Updated Gothic");
        updatedStyle.setId(id);
        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.empty());

        // when & then
        assertThrows(NotFoundException.class, () -> architecturalStyleService.update(id, updatedStyle));

        // verify
        verify(architecturalStyleRepository, times(1)).findById(id);
    }

    @Test
    void whenPatchArchitecturalStyle_thenPatchStyle() {
        // given
        int id = 1;
        ArchitecturalStyle existingStyle = new ArchitecturalStyle();
        existingStyle.setName("Gothic");
        existingStyle.setDescription("Old Description");
        existingStyle.setId(id);
        ArchitecturalStyle patchStyle = new ArchitecturalStyle();
        patchStyle.setName("Gothic");
        patchStyle.setDescription("New Description");
        patchStyle.setId(id);

        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.of(existingStyle));
        when(architecturalStyleRepository.save(existingStyle)).thenAnswer(invocation -> {
            existingStyle.setName(patchStyle.getName());
            return existingStyle;
        });

        // when
        ArchitecturalStyle result = architecturalStyleService.patchUpdate(id, patchStyle);

        // then
        verify(architecturalStyleRepository, times(1)).findById(id);
        verify(architecturalStyleRepository, times(1)).save(existingStyle);
        assertEquals(patchStyle.getName(), result.getName());
    }

    @Test
    void whenDeleteArchitecturalStyle_thenDeleteIfExists() {
        // given
        int id = 1;
        ArchitecturalStyle architecturalStyle = mock(ArchitecturalStyle.class);
        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.of(architecturalStyle));
        doNothing().when(architecturalStyleRepository).deleteById(id);

        // when
        architecturalStyleService.delete(id);

        // then
        verify(architecturalStyleRepository, times(1)).findById(id);
        verify(architecturalStyleRepository, times(1)).deleteById(id);
    }

    @Test
    void whenDeleteArchitecturalStyleAndStyleDoesNotExist_thenThrowNotFoundException() {
        // given
        int id = 1;
        when(architecturalStyleRepository.findById(id)).thenReturn(Optional.empty());

        // when & then
        assertThrows(NotFoundException.class, () -> architecturalStyleService.delete(id));

        // verify
        verify(architecturalStyleRepository, times(1)).findById(id);
        verify(architecturalStyleRepository, never()).deleteById(id);
    }

}

File: src/test/kotlin/by/architecture_map/belarus/service/impl/ArchitecturalStyleServiceImplTest.kt
package by.architecture_map.belarus.service.impl

import by.architecture_map.belarus.entity.ArchitecturalStyle
import by.architecture_map.belarus.repository.ArchitecturalStyleRepository
import by.architecture_map.belarus.repository.ImageRepository
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Test

import org.junit.jupiter.api.Assertions.*
import org.springframework.data.repository.findByIdOrNull
import java.util.*

class ArchitecturalStyleServiceImplTest {

    private val architecturalStyleRepository: ArchitecturalStyleRepository = mockk()
    private val imageRepository: ImageRepository = mockk()
    private val architecturalStyleService = ArchitecturalStyleServiceImpl(architecturalStyleRepository, imageRepository)

    @Test
    fun whenSaveArchitecturalStyle_thenReturnSavedStyle() {
        //given
        val architecturalStyle = ArchitecturalStyle(id = 1, name = "Gothic")
        every { architecturalStyleRepository.save(architecturalStyle) } returns architecturalStyle

        //when
        val result = architecturalStyleService.create(architecturalStyle)

        //then
        verify(exactly = 1) { architecturalStyleRepository.save(architecturalStyle) }
        assertEquals(architecturalStyle, result)
    }

    @Test
    fun whenFindAllArchitecturalStyles_thenReturnListOfStyles() {
        //given
        val styles = mutableListOf(
                ArchitecturalStyle(id = 1, name = "Gothic"),
                ArchitecturalStyle(id = 2, name = "Baroque")
        )
        every { architecturalStyleRepository.findAll() } returns styles

        //when
        val result = architecturalStyleService.findAll()

        //then
        verify(exactly = 1) { architecturalStyleRepository.findAll() }
        assertEquals(styles, result)
    }

    @Test
    fun whenFindArchitecturalStyleById_thenReturnStyle() {
        //given
        val id = 1
        val architecturalStyle = ArchitecturalStyle(id = id, name = "Gothic")
        every { architecturalStyleRepository.findByIdOrNull(id) } returns architecturalStyle

        //when
        val result = architecturalStyleService.findById(id)

        //then
        verify(exactly = 1) { architecturalStyleRepository.findByIdOrNull(id) }
        assertEquals(architecturalStyle, result)
    }

    @Test
    fun whenFindArchitecturalStyleByIdAndStyleDoesNotExist_thenReturnNull() {
        //given
        val id = 1
        every { architecturalStyleRepository.findByIdOrNull(id) } returns null

        //when
        val result = architecturalStyleService.findById(id)

        //then
        verify(exactly = 1) { architecturalStyleRepository.findByIdOrNull(id) }
        assertNull(result)
    }

    @Test
    fun whenUpdateArchitecturalStyleById_thenReturnUpdatedStyle() {
        //given
        val id = 1
        val existingStyle = ArchitecturalStyle(id = id, name = "Gothic")
        val updatedStyle = ArchitecturalStyle(id = id, name = "Updated Gothic")

        every { architecturalStyleRepository.findById(id) } returns Optional.of(existingStyle)
        every { architecturalStyleRepository.save(existingStyle) } returns existingStyle.apply {
            name = updatedStyle.name
        }

        //when
        val result = architecturalStyleService.updateById(id, updatedStyle)

        //then
        verify(exactly = 1) { architecturalStyleRepository.findById(id) }
        verify(exactly = 1) { architecturalStyleRepository.save(existingStyle) }
        assertEquals(updatedStyle.name, result?.name)
    }

    @Test
    fun whenUpdateArchitecturalStyleByIdAndStyleDoesNotExist_thenReturnNull() {
        //given
        val id = 1
        val updatedStyle = ArchitecturalStyle(id = id, name = "Updated Gothic")

        every { architecturalStyleRepository.findById(id) } returns Optional.empty()

        //when
        val result = architecturalStyleService.updateById(id, updatedStyle)

        //then
        verify(exactly = 1) { architecturalStyleRepository.findById(id) }
        assertNull(result)
    }

    @Test
    fun whenPatchArchitecturalStyleById_thenReturnPatchedStyle() {
        //given
        val id = 1
        val existingStyle = ArchitecturalStyle(id = id, name = "Gothic", description = "Old Description")
        val patchStyle = ArchitecturalStyle(id = id, description = "New Description")

        every { architecturalStyleRepository.findById(id) } returns Optional.of(existingStyle)
        every { architecturalStyleRepository.save(existingStyle) } returns existingStyle.apply {
            name = patchStyle.name
        }

        //when
        val result = architecturalStyleService.patchById(id, patchStyle)

        //then
        verify(exactly = 1) { architecturalStyleRepository.findById(id) }
        verify(exactly = 1) { architecturalStyleRepository.save(existingStyle) }
        assertEquals(patchStyle.name, result?.name)
    }

    @Test
    fun whenDeleteArchitecturalStyleById_thenReturnTrueIfExists() {
        //given
        val id = 1
        every { architecturalStyleRepository.existsById(id) } returns true
        every { architecturalStyleRepository.deleteById(id) } just Runs

        //when
        val result = architecturalStyleService.deleteById(id)

        //then
        verify(exactly = 1) { architecturalStyleRepository.existsById(id) }
        verify(exactly = 1) { architecturalStyleRepository.deleteById(id) }
        assertTrue(result)
    }

    @Test
    fun whenDeleteArchitecturalStyleByIdAndStyleDoesNotExist_thenReturnFalseIfNotExists() {
        //given
        val id = 1
        every { architecturalStyleRepository.existsById(id) } returns false
        every { architecturalStyleRepository.deleteById(id) } just Runs

        //when
        val result = architecturalStyleService.deleteById(id)

        //then
        verify(exactly = 1) { architecturalStyleRepository.existsById(id) }
        verify(exactly = 0) { architecturalStyleRepository.deleteById(id) }
        assertFalse(result)
    }
}